swap/peer.go

```diff
@@ -30,18 +35,144 @@ var ErrDontOwe = errors.New("no negative balance")
+	return p.lastReceivedCheque
+}
+
+func (p *Peer) getLastSentCheque() *Cheque {
+	return p.lastSentCheque
+}
+
+func (p *Peer) setLastReceivedCheque(cheque *Cheque) error {
+	p.lastReceivedCheque = cheque
+	return p.swap.saveLastReceivedCheque(p.ID(), cheque)
+}
+
+func (p *Peer) setLastSentCheque(cheque *Cheque) error {
+	p.lastSentCheque = cheque
+	return p.swap.saveLastSentCheque(p.ID(), cheque)
+}
+
+func (p *Peer) getLastChequeValues() (total uint64, err error) {
+	lastCheque := p.getLastReceivedCheque()
+	if lastCheque != nil {
+		total = lastCheque.CumulativePayout
+	}
+	return
+}
+
+func (p *Peer) setBalance(balance int64) error {
+	p.balance = balance
+	return p.swap.saveBalance(p.ID(), balance)
+}
+
+func (p *Peer) getBalance() int64 {
+	return p.balance
+}
+
+// To be called with mutex already held
+func (p *Peer) updateBalance(amount int64) error {
+	//adjust the balance
+	//if amount is negative, it will decrease, otherwise increase
+	newBalance := p.getBalance() + amount
+	if err := p.setBalance(newBalance); err != nil {
+		return err
+	}
+	log.Debug("balance for peer after accounting", "peer", p.ID().String(), "balance", strconv.FormatInt(newBalance, 10))
+	return nil
+}
+
+// createCheque creates a new cheque whose beneficiary will be the peer and
+// whose amount is based on the last cheque and current balance for this peer
+// The cheque will be signed and point to the issuer's contract
+// To be called with mutex already held
+// Caller must be careful that the same resources aren't concurrently read and written by multiple routines
+func (p *Peer) createCheque() (*Cheque, error) {
+	var cheque *Cheque
+	var err error
+
+	beneficiary := p.beneficiary
+	peerBalance := p.getBalance()
+	// the balance should be negative here, we take the absolute value:
+	honey := uint64(-peerBalance)
+	var amount uint64
+
+	// TODO: this must probably be locked
+	amount, err = p.swap.oracle.GetPrice(honey)
+	if err != nil {
+		return nil, fmt.Errorf("error getting price from oracle: %s", err.Error())
+	}
+
+	// if there is no existing cheque when loading from the store, it means it's the first interaction
+	// this is a valid scenario
+	total, err := p.getLastChequeValues()
+	if err != nil && err != state.ErrNotFound {
+		return nil, err
+	}
+
+	// TODO: lock
+	contract := p.swap.owner.Contract
+
+	cheque = &Cheque{
+		ChequeParams: ChequeParams{
+			CumulativePayout: total + amount,
+			Contract:         contract,
+			Beneficiary:      beneficiary,
+		},
+		Honey: honey,
+	}
+	cheque.Signature, err = cheque.Sign(p.swap.owner.privateKey)
+
+	return cheque, err
+}
+
+// sendCheque sends a cheque to peer
+// To be called with mutex already held
+// Caller must be careful that the same resources aren't concurrently read and written by multiple routines
+func (p *Peer) sendCheque() error {
+	cheque, err := p.createCheque()
+	if err != nil {
+		return fmt.Errorf("error while creating cheque: %s", err.Error())
```

Comment:
If returning the `err`, first evaluate:
* Has the error been logged before?
* Is the error text adding new additional information?
* If yes, use `fmt.Errorf("errortext %v", err)` instead of calling `err.Error()`
* If no, just return `err`

Reply:
Done.